src/polyhedralGravity/model/KDTree.cpp

```diff
@@ -0,0 +1,153 @@
+    }
+    return *this;
+}
+
+bool KDTree::operator==(const KDTree & other) const {
+    return this->rootNode == other.rootNode;
+}
+
+TreeNode &KDTree::getRootNode() {
+    if (!this->rootNode) {
+        this->rootNode = TreeNode::treeNodeFactory(*this->param.release());
+    }
+    return *this->rootNode;
+}
+
+unsigned long KDTree::countIntersections(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray) {
+    return this->getRootNode().countIntersections(origin, ray);
+}
+
+void KDTree::getFaceIntersections(const polyhedralGravity::Array3 &origin, const polyhedralGravity::Array3 &ray, std::vector<size_t> &intersectedFaceIndices) {
+    this->getRootNode().getFaceIntersections(origin, ray, intersectedFaceIndices);
+}
+
+
+std::tuple<Plane, double, TriangleIndexLists> KDTree::findPlane(const SplitParam &param) {// O(N^2) implementation
+    double cost = std::numeric_limits<double>::infinity();
+    Plane optPlane{0, param.splitDirection};
+    TriangleIndexLists optTriangleIndexLists{};
+    std::for_each(param.indexBoundFaces.cbegin(), param.indexBoundFaces.cend(), [&param, &optPlane, &cost, &optTriangleIndexLists](const size_t faceIndex) {
+        const auto &face = param.faces[faceIndex];
+        for (const auto &index: face) {
+            Plane candidatePlane{param.vertices[index][param.splitDirection], param.splitDirection};
+            if (auto [candidateCost, triangleIndexLists] = costForPlane(param, candidatePlane); candidateCost < cost) {
+                cost = candidateCost;
+                optPlane = candidatePlane;
+                optTriangleIndexLists = std::move(triangleIndexLists);
+            }
+        }
+    });
+    return std::make_tuple(std::move(optPlane), cost, std::move(optTriangleIndexLists));
+}
+
+Box KDTree::getBoundingBox(const std::vector<polyhedralGravity::Array3> &vertices) {
+    using namespace polyhedralGravity;
+    if(vertices.empty()) {
+        return {{0, 0, 0}, {0, 0, 0}};
+    }
+    Array3 min = vertices[0];
+    Array3 max = vertices[0];
+    std::for_each(vertices.cbegin(), vertices.cend(), [&min, &max](const Array3 &vertex) {
+        for (int i = 0; i < vertex.size(); i++) {
+            min[i] = std::min(min[i], vertex[i]);
+            max[i] = std::max(max[i], vertex[i]);
+        }
+    });
+    return {min, max};
+}
+
+std::pair<Box, Box> KDTree::splitBox(const Box &box, const Plane &plane) {
+    Box box1{box};
+    Box box2{box};
+    const Direction &axis{plane.second};
+    box1.second[axis] = plane.first;
+    box2.first[axis] = plane.first;
+    return std::make_pair(box1, box2);
+}
+
+std::pair<const double, TriangleIndexLists> KDTree::costForPlane(const SplitParam &param, const Plane &plane) {
+    //Checks if the split plane is one of the faces of the bounding box, if so the split is useless
+    if (plane.first == param.boundingBox.first[plane.second] || plane.first == param.boundingBox.second[plane.second]) {
```

Comment:
More line comments are needed for these more extensive methods. Rule of thumb: if it's longer than five lines ==> Document your actions. You don't need to document each line but can group similar functionality and have a blank line afterward. Have a look at `distancesToSegmentEndpoints` in `GravityModelDetail.cpp` as an example.

Reply:
General Comment ⬆️ This holds for multiple occasions in your code.

Reply:
I'll need to do an large overhaul then, I'll let you know when the code is more readable you can review again.